File: fetchr/run/scenario_outputs.py
from __future__ import annotations

from pathlib import Path
from typing import Any, Dict

import pandas as pd

from .artifact_schema import CURRENT_SCHEMA_VERSION
from .json_utils import write_json


def _read_date_indexed_csv(path: Path) -> pd.DataFrame:
    frame = pd.read_csv(path)
    if "date" not in frame.columns:
        raise ValueError(f"Scenario artifact is missing date column: {path}")
    frame["date"] = pd.to_datetime(frame["date"], errors="coerce")
    frame = frame.dropna(subset=["date"]).set_index("date").sort_index()
    frame = frame[~frame.index.duplicated(keep="last")]
    return frame


def _quarterly_sparse_from_monthly(series: pd.Series, agg: str = "last") -> pd.Series:
    s = pd.to_numeric(series, errors="coerce")
    s.index = pd.to_datetime(s.index)
    s = s.sort_index()
    q = s.groupby(s.index.to_period("Q"))
    if agg == "sum":
        qvals = q.sum(min_count=1)
    elif agg == "mean":
        qvals = q.mean()
    elif agg == "first":
        qvals = q.first()
    else:
        qvals = q.last()
    out = pd.Series(index=s.index, dtype=float, name=str(series.name or "series"))
    q_end_idx = qvals.index.to_timestamp(how="end").normalize()
    out.loc[q_end_idx] = qvals.values
    return out


def build_scenario_outputs(cfg: Dict[str, Any], interpolation_summary: pd.DataFrame) -> Dict[str, Any]:
    scenario_dir = Path(cfg["SCENARIO_DIR"])
    quant_dir = scenario_dir / "quantiles"
    rep_dir = scenario_dir / "representatives"
    quant_dir.mkdir(parents=True, exist_ok=True)
    rep_dir.mkdir(parents=True, exist_ok=True)

    summary: Dict[str, Any] = {
        "schema_version": CURRENT_SCHEMA_VERSION,
        "n_dfm_tasks": 0,
        "n_quantile_files": 0,
        "n_representative_files": 0,
        "n_mixed_quantile_panels": 0,
        "tasks": [],
    }

    if interpolation_summary.empty:
        write_json(Path(cfg["SCENARIO_SUMMARY_JSON"]), summary)
        return summary

    quantile_series_by_label: dict[str, dict[str, pd.Series]] = {}

    for _, row in interpolation_summary.iterrows():
        if str(row.get("status", "")).strip().lower() != "ok":
            continue
        if str(row.get("method", "")).strip().lower() != "quarterly_to_monthly_dfm_state_space":
            continue

        task_name = str(row.get("name", "")).strip()
        artifact_dir_text = str(row.get("artifact_dir", "")).strip()
        if not task_name or not artifact_dir_text:
            continue
        artifact_dir = Path(artifact_dir_text)
        if not artifact_dir.exists():
            continue

        summary["n_dfm_tasks"] = int(summary["n_dfm_tasks"]) + 1
        task_info: Dict[str, Any] = {
            "task_name": task_name,
            "artifact_dir": str(artifact_dir),
            "quantiles_csv": "",
            "representatives_csv": "",
        }

        quantiles_path = artifact_dir / "bootstrap_quantiles.csv"
        if quantiles_path.exists():
            qdf = _read_date_indexed_csv(quantiles_path)
            qdf_out = quant_dir / f"{task_name}_quantiles.csv"
            qdf.to_csv(qdf_out, index_label="date")
            task_info["quantiles_csv"] = str(qdf_out)
            summary["n_quantile_files"] = int(summary["n_quantile_files"]) + 1

            for col in qdf.columns:
                label = str(col).strip()
                if not label:
                    continue
                quantile_series_by_label.setdefault(label, {})[task_name] = pd.to_numeric(
                    qdf[col], errors="coerce"
                ).rename(task_name)

        representatives_path = artifact_dir / "bootstrap_representative_paths.csv"
        if representatives_path.exists():
            rdf = _read_date_indexed_csv(representatives_path)
            rdf_out = rep_dir / f"{task_name}_representatives.csv"
            rdf.to_csv(rdf_out, index_label="date")
            task_info["representatives_csv"] = str(rdf_out)
            summary["n_representative_files"] = int(summary["n_representative_files"]) + 1

        summary["tasks"].append(task_info)

    for quantile_label, series_map in quantile_series_by_label.items():
        if not series_map:
            continue
        dense = pd.concat(series_map.values(), axis=1).sort_index()
        dense = dense[~dense.index.duplicated(keep="last")]
        sparse_map = {
            name: _quarterly_sparse_from_monthly(s, agg="last")
            for name, s in series_map.items()
        }
        sparse = pd.concat(sparse_map.values(), axis=1).reindex(dense.index)
        sparse.columns = list(dense.columns)

        dense_out = scenario_dir / f"mixed_{quantile_label}_dense.csv"
        sparse_out = scenario_dir / f"mixed_{quantile_label}_sparse.csv"
        dense.to_csv(dense_out, index_label="date")
        sparse.to_csv(sparse_out, index_label="date")
        summary["n_mixed_quantile_panels"] = int(summary["n_mixed_quantile_panels"]) + 1

    write_json(Path(cfg["SCENARIO_SUMMARY_JSON"]), summary)
    return summary

File: econark-r/fetchr-R/run/scenario_outputs.R
.read_date_indexed_csv <- function(path) {
  df <- utils::read.csv(path, stringsAsFactors = FALSE, check.names = FALSE)
  if (!"date" %in% names(df)) {
    stop(sprintf("Scenario artifact is missing date column: %s", path))
  }
  df$date <- as.Date(df$date)
  df <- df[!is.na(df$date), , drop = FALSE]
  df <- df[order(df$date), , drop = FALSE]
  df <- df[!duplicated(df$date, fromLast = TRUE), , drop = FALSE]
  for (cc in setdiff(names(df), "date")) {
    df[[cc]] <- suppressWarnings(as.numeric(df[[cc]]))
  }
  rownames(df) <- NULL
  df
}

.merge_quantile_series <- function(series_map) {
  names_vec <- names(series_map)
  if (length(names_vec) == 0L) return(data.frame(date = as.Date(character())))

  parts <- list()
  for (nm in names_vec) {
    src <- series_map[[nm]]
    part <- data.frame(
      date = as.Date(src$date),
      value = suppressWarnings(as.numeric(src$value)),
      stringsAsFactors = FALSE
    )
    names(part)[2] <- nm
    parts[[length(parts) + 1L]] <- part
  }

  out <- parts[[1L]]
  if (length(parts) > 1L) {
    for (i in 2:length(parts)) out <- merge(out, parts[[i]], by = "date", all = TRUE)
  }
  out <- out[order(out$date), , drop = FALSE]
  out <- out[!duplicated(out$date, fromLast = TRUE), , drop = FALSE]
  rownames(out) <- NULL
  out
}

build_scenario_outputs <- function(cfg, interpolation_summary) {
  scenario_dir <- cfg$SCENARIO_DIR
  quant_dir <- file.path(scenario_dir, "quantiles")
  rep_dir <- file.path(scenario_dir, "representatives")
  dir.create(scenario_dir, recursive = TRUE, showWarnings = FALSE)
  dir.create(quant_dir, recursive = TRUE, showWarnings = FALSE)
  dir.create(rep_dir, recursive = TRUE, showWarnings = FALSE)

  summary <- list(
    schema_version = 1L,
    n_dfm_tasks = 0L,
    n_quantile_files = 0L,
    n_representative_files = 0L,
    n_mixed_quantile_panels = 0L,
    tasks = list()
  )

  if (!is.data.frame(interpolation_summary) || nrow(interpolation_summary) == 0L || !"artifact_dir" %in% names(interpolation_summary)) {
    write_json_file(cfg$SCENARIO_SUMMARY_JSON, summary)
    return(invisible(summary))
  }

  quantile_series_by_label <- list()

  for (i in seq_len(nrow(interpolation_summary))) {
    status <- tolower(trimws(as.character(interpolation_summary$status[[i]])))
    method <- tolower(trimws(as.character(interpolation_summary$method[[i]])))
    if (status != "ok" || method != "quarterly_to_monthly_dfm_state_space") next

    task_name <- trimws(as.character(interpolation_summary$name[[i]]))
    artifact_dir <- trimws(as.character(interpolation_summary$artifact_dir[[i]]))
    if (!nzchar(task_name) || !nzchar(artifact_dir) || !dir.exists(artifact_dir)) next

    summary$n_dfm_tasks <- as.integer(summary$n_dfm_tasks) + 1L
    task_info <- list(
      task_name = task_name,
      artifact_dir = artifact_dir,
      quantiles_csv = "",
      representatives_csv = ""
    )

    quantiles_path <- file.path(artifact_dir, "bootstrap_quantiles.csv")
    if (file.exists(quantiles_path)) {
      qdf <- .read_date_indexed_csv(quantiles_path)
      qdf_out <- file.path(quant_dir, paste0(task_name, "_quantiles.csv"))
      utils::write.csv(qdf, qdf_out, row.names = FALSE)
      task_info$quantiles_csv <- qdf_out
      summary$n_quantile_files <- as.integer(summary$n_quantile_files) + 1L

      for (cc in setdiff(names(qdf), "date")) {
        label <- trimws(as.character(cc))
        if (!nzchar(label)) next
        if (is.null(quantile_series_by_label[[label]])) quantile_series_by_label[[label]] <- list()
        quantile_series_by_label[[label]][[task_name]] <- data.frame(
          date = qdf$date,
          value = suppressWarnings(as.numeric(qdf[[cc]])),
          stringsAsFactors = FALSE
        )
      }
    }

    reps_path <- file.path(artifact_dir, "bootstrap_representative_paths.csv")
    if (file.exists(reps_path)) {
      rdf <- .read_date_indexed_csv(reps_path)
      rdf_out <- file.path(rep_dir, paste0(task_name, "_representatives.csv"))
      utils::write.csv(rdf, rdf_out, row.names = FALSE)
      task_info$representatives_csv <- rdf_out
      summary$n_representative_files <- as.integer(summary$n_representative_files) + 1L
    }

    summary$tasks[[length(summary$tasks) + 1L]] <- task_info
  }

  for (label in names(quantile_series_by_label)) {
    series_map <- quantile_series_by_label[[label]]
    if (length(series_map) == 0L) next

    dense <- .merge_quantile_series(series_map)
    sparse <- dense
    if (ncol(sparse) > 1L) {
      keep <- as.integer(format(sparse$date, "%m")) %in% c(3L, 6L, 9L, 12L)
      for (cc in setdiff(names(sparse), "date")) {
        sparse[[cc]][!keep] <- NA_real_
      }
    }

    dense_out <- file.path(scenario_dir, paste0("mixed_", label, "_dense.csv"))
    sparse_out <- file.path(scenario_dir, paste0("mixed_", label, "_sparse.csv"))
    utils::write.csv(dense, dense_out, row.names = FALSE)
    utils::write.csv(sparse, sparse_out, row.names = FALSE)
    summary$n_mixed_quantile_panels <- as.integer(summary$n_mixed_quantile_panels) + 1L
  }

  write_json_file(cfg$SCENARIO_SUMMARY_JSON, summary)
  invisible(summary)
}
